### PR TITLE
Refactor assignment warnings

### PR DESCRIPTION
**Pull request by jcouv**
From discussion with @cston, there is more work to do to factor the logic and checks whenever assigning/converting a value (assignment, local declaration, return, yield return, argument passing, foreach variable initialization, ...), so that it can easily be re-used.
This PR is a small step in that direction.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -482,36 +482,87 @@ private static object GetTypeAsDiagnosticArgument(TypeSymbol typeOpt)
...
+                else
+                {
+                    ReportDiagnostic(assignmentKind == AssignmentKind.Return ? ErrorCode.WRN_NullReferenceReturn : ErrorCode.WRN_NullAsNonNullable, value.Syntax);
+                }
+                return true;
+            }
         }
 
-        private void ReportAssignmentWarnings(BoundExpression value, TypeSymbolWithAnnotations targetType, TypeSymbolWithAnnotations valueType, bool useLegacyWarnings)
+        private void ReportAssignmentWarnings(BoundExpression value, TypeSymbolWithAnnotations targetType, TypeSymbolWithAnnotations valueType, bool useLegacyWarnings, AssignmentKind assignmentKind = AssignmentKind.Assignment)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >AssignmentKind assignmentKind = AssignmentKind.Assignment [](start = 168, length = 57)

Do any callers pass this parameter explicitly?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -523,9 +574,23 @@ private void ReportAssignmentWarnings(BoundExpression value, TypeSymbolWithAnnot
                     return;
                 }
 
-                ReportNullReferenceAssignmentIfNecessary(value, targetType, valueType, useLegacyWarnings);
-                ReportNullabilityMismatchInAssignmentIfNecessary(value, valueType.TypeSymbol, targetType.TypeSymbol);
+                ReportNullableAssignmentIfNecessary(value, targetType, valueType, useLegacyWarnings, assignmentKind);
+                reportNullabilityMismatchInAssignmentIfNecessary();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >reportNullabilityMismatchInAssignmentIfNecessary [](start = 16, length = 48)

Consider inlining.

**Comment by jcouv** [comment]
Hum, looks like I broke bootstrap build. Investigating.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -3410,20 +3445,7 @@ public override BoundNode VisitAddressOfOperator(BoundAddressOfOperator node)
...
-                    {
-                        ReportDiagnostic(ErrorCode.WRN_NullReferenceArgument, argument.Syntax,
-                            new FormattedSymbol(parameter, SymbolDisplayFormat.ShortFormat),
-                            new FormattedSymbol(parameter.ContainingSymbol, SymbolDisplayFormat.MinimallyQualifiedFormat));
-                    }
-                    return true;
-                }
-            }
-            return false;
+            return ReportNullableAssignmentIfNecessary(argument, paramType, argumentType, useLegacyWarnings: false, assignmentKind: AssignmentKind.Argument, target: parameter);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Found the issue: the handling of arguments in object creation is different than that in calls. The former keeps conversion, while the latter strips them.
The issue is that a conversion on a argument that produces a split state is unexpected. Previously, `ReportNullReferenceArgumentIfNecessary` didn't assert an unsplit state, but re-using `ReportNullableAssignmentIfNecessary` means it now does...

Chuck, we can discuss what is the correct way of handling this in the new model for handling assignments and conversions.

Here's an example:
```C#
        [Fact]
        public void TestUnaryNegation()
        {
            var source = @"
public class C<T>
{
    C(C<object> c) => throw null;
    void M(bool b)
    {
        _ = new C<object>(!b);
    }
    public static implicit operator C<T>(T x) => throw null;
}
";
 
            var c = CreateCompilation(source, parseOptions: TestOptions.Regular8);
            c.VerifyDiagnostics( );
        }
```

```
   at System.Diagnostics.Debug.Assert(Boolean condition)
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.ReportNullableAssignmentIfNecessary(BoundExpression value, TypeSymbolWithAnnotations targetType, TypeSymbolWithAnnotations valueType, Boolean useLegacyWarnings, AssignmentKind assignmentKind, Symbol target) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 497
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.ReportNullReferenceArgumentIfNecessary(BoundExpression argument, TypeSymbolWithAnnotations argumentType, ParameterSymbol parameter, TypeSymbolWithAnnotations paramType) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 3448
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.ApplyConversion(BoundNode node, BoundExpression operandOpt, Conversion conversion, TypeSymbol targetType, TypeSymbolWithAnnotations operandType, Boolean checkConversion, Boolean fromExplicitCast, Boolean& canConvertNestedNullability) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 3031
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitConversion(BoundConversion node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2801
   at Microsoft.CodeAnalysis.CSharp.BoundConversion.Accept(BoundTreeVisitor visitor) in C:\repos\roslyn\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 2091
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.Visit(BoundNode node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 147
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitExpressionWithoutStackGuard(BoundExpression node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\PreciseAbstractFlowPass.cs:line 304
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitExpressionWithoutStackGuard(BoundExpression node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 1046
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 202
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitWithStackGuard(BoundNode node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\PreciseAbstractFlowPass.cs:line 296
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitAlways(BoundNode node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\PreciseAbstractFlowPass.cs:line 284
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.Visit(BoundNode node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\PreciseAbstractFlowPass.cs:line 266
   at Microsoft.CodeAnalysis.CSharp.PreciseAbstractFlowPass`1.VisitRvalue(BoundExpression node) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\PreciseAbstractFlowPass.cs:line 611
   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitArgumentEvaluate(ImmutableArray`1 arguments, ImmutableArray`1 refKindsOpt, Int32 i, Boolean preserveConditionalState) in C:\repos\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2164
```

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 4]
```diff
@@ -494,7 +494,8 @@ private enum AssignmentKind
         /// </summary>
         private bool ReportNullableAssignmentIfNecessary(BoundExpression value, TypeSymbolWithAnnotations targetType, TypeSymbolWithAnnotations valueType, bool useLegacyWarnings, AssignmentKind assignmentKind = AssignmentKind.Assignment, Symbol target = null)
         {
-            Debug.Assert(!IsConditionalState);
+            // PROTOTYPE(NullableReferenceTypes): Need to discuss
+            //Debug.Assert(!IsConditionalState);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: This assert can be removed. Reporting warnings does not depend on whether the state is conditional.

**Comment by jcouv** [comment]
@AlekseyTs @333fred @dotnet/roslyn-compiler for a second review. Thanks

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 5]
```diff
@@ -3453,20 +3488,7 @@ public override BoundNode VisitAddressOfOperator(BoundAddressOfOperator node)
         /// </summary>
         private bool ReportNullReferenceArgumentIfNecessary(BoundExpression argument, TypeSymbolWithAnnotations argumentType, ParameterSymbol parameter, TypeSymbolWithAnnotations paramType)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Should we consider removing this method altogether and having callers just call the underlying version?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 6]
```diff
@@ -19056,9 +19075,9 @@ static void Test2()
                 // (21,26): warning CS8600: Converting null literal or possible null value to non-nullable type.
                 //                     p2 = null;
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Nit: Update warning comments.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 7]
```diff
@@ -49,6 +49,25 @@ static void Main()
                 );
         }
 
+        [Fact]
+        public void TestUnaryNegation()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: If this is testing something nullablewalker, it should have the `NonNullTypes` attribute.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 5 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: I'll leave that for a follow-up (Chuck is doing further simplifications).

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 6 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: It seems to be up-to-date already.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 7 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: In DEBUG mode, we always run NullableWalker (regardless of NonNullTypes context). That's why that test was still useful to hit an assertion (now removed).

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 6 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: @jcouv as in there's 3 warnings but only 2 comments in the code about the warnings.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 7 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: @jcouv that means this is relying on a side effect to be a valid test, not actually doing anything useful directly.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 6 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Sorry, I'd misunderstood. I'll add that to upcoming PR. Thanks

**Pull request merged**